src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php

```diff
@@ -86,77 +94,92 @@ public function close(): bool
+        );
+
+        $this->manager->executeBulkWrite($this->namespace, $write);
 
         return true;
     }
 
     public function gc(int $maxlifetime): int|false
     {
-        return $this->getCollection()->deleteMany([
-            $this->options['expiry_field'] => ['$lt' => new UTCDateTime()],
-        ])->getDeletedCount();
+        $write = new BulkWrite();
+        $write->delete(
+            [$this->options['expiry_field'] => ['$lt' => $this->getUTCDateTime()]],
+        );
+        $result = $this->manager->executeBulkWrite($this->namespace, $write);
+
+        return $result->getDeletedCount() ?? false;
     }
 
     protected function doWrite(#[\SensitiveParameter] string $sessionId, string $data): bool
     {
         $ttl = ($this->ttl instanceof \Closure ? ($this->ttl)() : $this->ttl) ?? \ini_get('session.gc_maxlifetime');
-        $expiry = new UTCDateTime((time() + (int) $ttl) * 1000);
+        $expiry = $this->getUTCDateTime($ttl);
 
         $fields = [
-            $this->options['time_field'] => new UTCDateTime(),
+            $this->options['time_field'] => $this->getUTCDateTime(),
             $this->options['expiry_field'] => $expiry,
             $this->options['data_field'] => new Binary($data, Binary::TYPE_OLD_BINARY),
         ];
 
-        $this->getCollection()->updateOne(
+        $write = new BulkWrite();
+        $write->update(
             [$this->options['id_field'] => $sessionId],
             ['$set' => $fields],
             ['upsert' => true]
         );
 
+        $this->manager->executeBulkWrite($this->namespace, $write);
+
         return true;
     }
 
     public function updateTimestamp(#[\SensitiveParameter] string $sessionId, string $data): bool
     {
         $ttl = ($this->ttl instanceof \Closure ? ($this->ttl)() : $this->ttl) ?? \ini_get('session.gc_maxlifetime');
-        $expiry = new UTCDateTime((time() + (int) $ttl) * 1000);
+        $expiry = $this->getUTCDateTime($ttl);
 
-        $this->getCollection()->updateOne(
+        $write = new BulkWrite();
+        $write->update(
             [$this->options['id_field'] => $sessionId],
             ['$set' => [
-                $this->options['time_field'] => new UTCDateTime(),
+                $this->options['time_field'] => $this->getUTCDateTime(),
                 $this->options['expiry_field'] => $expiry,
-            ]]
+            ]],
+            ['limit' => 1],
         );
 
+        $this->manager->executeBulkWrite($this->namespace, $write);
+
         return true;
     }
 
     protected function doRead(#[\SensitiveParameter] string $sessionId): string
     {
-        $dbData = $this->getCollection()->findOne([
+        $dbData = $this->manager->executeQuery($this->namespace, new Query([
             $this->options['id_field'] => $sessionId,
-            $this->options['expiry_field'] => ['$gte' => new UTCDateTime()],
-        ]);
+            $this->options['expiry_field'] => ['$gte' => $this->getUTCDateTime()],
+        ]));
```

Comment:
Since we're only working with the data field, it might be sensible to also add a projection to the query:

```suggestion
            $this->options['expiry_field'] => ['$gte' => $this->getUTCDateTime()],
        ], [
            'projection' => [$this->options['data_field'] => true]
        ]));
```

While the metadata is not that big, reducing the result document to just binary data would almost completely eliminate the overhead of using the default driver type map, allowing you to skip working with raw BSON and thus changing the driver version requirement.

(Apologies for the suboptimal formatting, but GitHub wouldn't let me add a suggestion for the whole block due to the "deleted" line).

Reply:
💡 Very smart.